src/client/opamCommands.ml

```diff
@@ -2860,6 +2860,14 @@ let pin ?(unpin_only=false) () =
     mk_flag ["dev-repo"] "Pin to the upstream package source for the latest \
                           development version"
   in
+  let with_version =
+    mk_opt ["with-version"] "VERSION"
+      "Set $(b,PACKAGES) version to $(b,VERSION). If $(b,PACKAGES) contains a \
+       version, it is ignored. With version pin, pinned package source are \
+       taken from the version pin and the package version is the specified \
+       with this option."
+      Arg.(some package_version) None
```

Comment:
This does not make it totally clear to me:
- does it force the version of all pinnings ?
- does is affect the source used for pinning, or the version it is pinned as, or both ?
- how does it interact with `opam pin foo.VERSION` ? I guess it should have lower priority or be an error, unless their meaning are different ; it seems to me this option has the same purpose, but the point is to scale with pinning URLs directly

Note that at the moment, weird stuff like `opam pin foo.2 3` is allowed, and simply ignores the `2`, so it's a version pin similar to `opam pin foo 3`. Maybe it should be an error ? You _could_ want to pin the package definition of version `2`, while pretending it's version `3` though ; that would no longer qualify as a version pinning however.

Reply:
>  does it force the version of all pinnings ?

yes all pinned packages of the current command

>    does is affect the source used for pinning, or the version it is pinned as, or both ?

It affects only the version of the pinned packages, it is similar to `opam pin <whatever sort of pinning> --edit` with a version change at edit stage.

>  version clash

In this implementation, the `--with-version` version is the one taken (highest priority), regardless of `foo.VERSION`, `foo version`, etc. On `opam pin foo 2 --with-version 3`, the code of version `2` is taken, and the package pinned with version `3` (cf. new [Source_version](https://github.com/ocaml/opam/pull/4301/files#diff-fbfddeb93a7d76d66f9bfd9b6cfbfec1R3096)).

Reply:
I am interested in the answers, but my point was that they should be more obvious from the commands/doc ;)
As for `opam pin foo.V1 --with-version V2`, it would seem more obvious to me to choose `V1`, since it's syntactlically "more specific" than `V2`. Of course, with a single package pinning, this makes no sense, and it's the main case... so maybe an error is still best.

Reply:
Funny, I had another (obvious) thought process about `opam pin foo.V1 --with-version V2`: as user specified the version with an additional version, `V2` has priority.
For the doc :+1:

Reply:
Updated